service/src/main/java/bio/terra/pipelines/service/JobsService.java

```diff
@@ -81,18 +87,32 @@ protected UUID writeJobToDb(
     }
   }
 
+  protected UUID writeJobToDbRetryDuplicateException(DbJob dbJob) {
+    Optional<DbJob> jobExists = jobsRepository.findJobByJobId(dbJob.getJobId());
```

Comment:
`save()` does not return duplicate key constraints for `@Id`annotated columns (at least by default and is not the easiest thing to configure to do so, the next sentence is a better solution in my opinion along with just better practice as you should never really share PKs of tables externally if possible).  It will actually just update the existing row than try to make a new row.  I want to change the table to actually have a new `id` column and make that the `@Id` annotated column with a unique constraint on `job_id` so I wouldn't haven to do this at all but again trying to keep the changes to a minimum.  

This is not wrapped in a transaction and I'm not entirely sure it needs to be.  Would love to discuss further (along with the fact that my final plan is to not need this call at all and just have a `save()` that can throw a unique key constraint

Reply:
> save() does not return duplicate key constraints for @Idannotated columns

Oh, I assumed that if it was an `@Id` it would have a unique key constraint in the db. I think JPA assumes the ID field is unique, I'm not sure what would happen if it wasn't.

> I want to change the table to actually have a new id column and make that the @Id annotated column with a unique constraint on job_id so I wouldn't haven to do this at all but again trying to keep the changes to a minimum.

If the db table is still in flux then maybe it's not worth trying to get this all correct right now.

> This is not wrapped in a transaction and I'm not entirely sure it needs to be

My mental model may be off, but I think about it like a concurrency problem. So if your logic can be broken by some db operation occurring in the middle, then you want to create a transaction so the operation is retried if the db changes out from under it.

Reply:
Regarding transactions if you are using terra-common-lib then you can us `@ReadTransaction` or `@WriteTransaction`. See https://github.com/DataBiosphere/terra-common-lib/tree/develop/src/main/java/bio/terra/common/db

Reply:
>Oh, I assumed that if it was an @Id it would have a unique key constraint in the db. I think JPA assumes the ID field is unique, I'm not sure what would happen if it wasn't.

it is expected to be unique but by default at a high level the way it works is that if the @Id field is populated, if that row already exists in the table then it will just do an update, if it doesnt exist it will create new row.  If the @Id column is `null` it will also try to create a new row (usually you have an autogenerated column as the `@Id` so it generates a value for you).  This is why I want to add a new autogenerated `id` column and make it the `@Id`  since right now we are populating the job_id ourselves and that doesn't really make sense with the way `CrudRepository` works by default.

With the right table structure (@Id on the new `id` column and unique constraint on `job_id`) things would behave a lot more like someone would expect.  I want to make a PR to make changes to all the tables and clean up the migrations to me minimal so that we can drop our database in dev and start fresh as opposed to making a bunch of mini migrations in each pr and having it look a little gross at the jump.  Since this project is so new might as well take advantage of it